README.md
# SageMaker Containers

SageMaker Containers contains common functionality necessary to create a container compatible with SageMaker. It can be simply used by any container by just installing the module:

```bash
pip install sagemaker-containers
```

SageMaker Containers gives you tools to create SageMaker-compatible containers, and has additional tools for letting you create Frameworks (SageMaker-compatible containers that can run arbitrary python scripts).

[SageMaker Chainer Container](https://github.com/aws/sagemaker-chainer-container) uses sagemaker-containers.

[SageMaker TensorFlow Container](https://github.com/aws/sagemaker-tensorflow-containers) and [SageMaker MXNet Container](https://github.com/aws/sagemaker-mxnet-containers) will be ported to use it as well in the future. 

## Getting Started -  Executing User Scripts on Amazon SageMaker

The objective of this tutorial is to explain how a script is executed inside any SageMaker-compatible container using **SageMaker Containers**.

### Creating the training job

A SageMaker training job created using the [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk#sagemaker-python-sdk-overview) [```Chainer```](https://github.com/aws/sagemaker-python-sdk#chainer-sagemaker-estimators), [```TensorFlow```](https://github.com/aws/sagemaker-python-sdk#tensorflow-sagemaker-estimators) and [```MXNet```](https://github.com/aws/sagemaker-python-sdk#mxnet-sagemaker-estimators) takes an user script containing the model to be trained, the Hyperparameters required by the script, and information about the input data. For example:

```python
from sagemaker.chainer import Chainer

# for complete list of parameters, see 
# https://github.com/aws/sagemaker-python-sdk#sagemaker-python-sdk-overview
estimator = Chainer(entry_point='user-script.py', 
                    hyperparameters={'batch-size':256, 
                                     'learning-rate':0.0001, 
                                     'communicator':'pure_nccl'},
                    ...) 

# starts the training job with an input data channel named training pointing to 
# s3://bucket/path/to/training/data
# for more information about data channels, see
# https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html#your-algorithms-training-algo-running-container-inputdataconfig
chainer_estimator.fit({'training': 's3://bucket/path/to/training/data', 'testing': 's3://bucket/path/to/testing/data')
```

### How a script is executed inside the container

When the container starts for training, **SageMaker Containers** installs the user script as a Python module. The module name matches the script name. In the case above, **user-script.py** is transformed in a Python module named **user-script**.

After that, the Python interpreter executes the user module, passing ```hyperparameters``` as script arguments. The example above will be executed by **SageMaker Containers** as follow:

```bash
python -m user-script --batch-size 256 --learning_rate 0.0001 --communicator pure_nccl
```

A user provide script consumes the hyperparameters using any argument parsing library, for example:

```python
if __name__ == '__main__':
  parser = argparse.ArgumentParser()

  parser.add_argument('--learning-rate', type=int, default=1)
  parser.add_argument('--batch-size', type=int, default=64)
  parser.add_argument('--communicator', type=str)
  parser.add_argument('--frequency', type=int, default=20)

  args = parser.parse_args()
  ...
```

### Reading additional information from the container

Very often, a user script needs additional information from the container that is not available in ```hyperparameters```.
SageMaker Containers writes this information as **environment variables** that are available inside the script.

For example, the example above can read information about the **training** channel provided in the training job request:

```python
if __name__ == '__main__':
  parser = argparse.ArgumentParser()

  ...
  
  # reads input channels training and testing from the environment variables
  parser.add_argument('--training', type=str, default=os.environ['SM_CHANNEL_TRAINING'])
  parser.add_argument('--testing', type=str, default=os.environ['SM_CHANNEL_TESTING'])

  args = parser.parse_args()
  ...
```
### List of provided environment variables by SageMaker Containers

The list of the environment variables is logged and available in cloudwatch logs. From the example above:
```bash
SM_NUM_GPUS=1
SM_NUM_CPUS=4
SM_NETWORK_INTERFACE_NAME=ethwe

SM_CURRENT_HOST=algo-1
SM_HOSTS=["algo-1","algo-2"]
SM_LOG_LEVEL=20

SM_USER_ARGS=["--batch-size","256","--learning-rate","0.0001","--communicator","pure_nccl"]

SM_HP_LEARNING_RATE=0.0001
SM_HP_BATCH-SIZE=10000

SM_HPS={"batch-size": '256', "learning-rate": "0.0001","communicator": "pure_nccl"}

SM_CHANNELS=["testing","training"]
SM_CHANNEL_TRAINING=/opt/ml/input/data/training
SM_CHANNEL_TESTING=/opt/ml/input/data/test

SM_MODULE_NAME=user_script
SM_MODULE_DIR=s3://sagemaker-{aws-region}-{aws-id}/{training-job-name}/source/sourcedir.tar.gz

SM_INPUT_DIR=/opt/ml/input
SM_INPUT_CONFIG_DIR=/opt/ml/input/config
SM_OUTPUT_DIR=/opt/ml/output
SM_OUTPUT_DATA_DIR=/opt/ml/output/data/algo-1
SM_MODEL_DIR=/opt/ml/model

SM_RESOURCE_CONFIG=
{
    "current_host": "algo-1",
    "hosts": [
        "algo-1",
        "algo-2"
    ]
}

SM_INPUT_DATA_CONFIG=
{
    "test": {
        "RecordWrapperType": "None",
        "S3DistributionType": "FullyReplicated",
        "TrainingInputMode": "File"
    },
    "train": {
        "RecordWrapperType": "None",
        "S3DistributionType": "FullyReplicated",
        "TrainingInputMode": "File"
    }
}


SM_FRAMEWORK_MODULE=sagemaker_chainer_container.training:main

SM_TRAINING_ENV=
{
    "channel_input_dirs": {
        "test": "/opt/ml/input/data/testing",
        "train": "/opt/ml/input/data/training"
    },
    "current_host": "algo-1",
    "framework_module": "sagemaker_chainer_container.training:main",
    "hosts": [
        "algo-1",
        "algo-2"
    ],
    "hyperparameters": {
        "batch-size": 10000,
        "epochs": 1
    },
    "input_config_dir": "/opt/ml/input/config",
    "input_data_config": {
        "test": {
            "RecordWrapperType": "None",
            "S3DistributionType": "FullyReplicated",
            "TrainingInputMode": "File"
        },
        "train": {
            "RecordWrapperType": "None",
            "S3DistributionType": "FullyReplicated",
            "TrainingInputMode": "File"
        }
    },
    "input_dir": "/opt/ml/input",
    "job_name": "preprod-chainer-2018-05-31-06-27-15-511",
    "log_level": 20,
    "model_dir": "/opt/ml/model",
    "module_dir": "s3://sagemaker-{aws-region}-{aws-id}/{training-job-name}/source/sourcedir.tar.gz",
    "module_name": "user_script",
    "network_interface_name": "ethwe",
    "num_cpus": 4,
    "num_gpus": 1,
    "output_data_dir": "/opt/ml/output/data/algo-1",
    "output_dir": "/opt/ml/output",
    "resource_config": {
        "current_host": "algo-1",
        "hosts": [
            "algo-1",
            "algo-2"
        ]
    }
}
```
## IMPORTANT ENVIRONMENT VARIABLES
These environment variables are those that you're likely to use when writing a user script. A full list of environment variables is given below.
### SM_MODEL_DIR
```json
SM_MODEL_DIR=/opt/ml/model
```
When the training job finishes, the container will be **deleted** including its file system expect for **/opt/ml/model** and **/opt/ml/output**. Use **/opt/ml/model** to save the model checkpoints. These checkpoints will be uploaded to the default S3 bucket. Usage example:
```python
# using it in argparse
parser.add_argument('model_dir', type=str, default=os.environ['SM_MODEL_DIR'])

# using it as variable
model_dir = os.environ['SM_MODEL_DIR']

# saving checkpoints to model dir in chainer
serializers.save_npz(os.path.join(os.environ['SM_MODEL_DIR'], 'model.npz'), model)
```

For more information, see: [How Amazon SageMaker Processes Training Output](https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html#your-algorithms-training-algo-envvariables).

### SM_CHANNELS
```bash
SM_CHANNELS='["testing","training"]'
```
Contains the list of input data channels in the container.

When you run training, you can partition your training data into different logical "channels".
Depending on your problem, some common channel ideas are: "training", "testing", "evaluation" or "images'and "labels".

```SM_CHANNELS``` includes the name of the available channels in the container as a JSON encoded list. Usage example:

```python
import json

# using it in argparse
parser.add_argument('channel_names', type=int, default=json.loads(os.environ['SM_CHANNELS'])))

# using it as variable
channel_names = json.loads(os.environ['SM_CHANNELS']))
```

### SM_CHANNEL_```{channel_name}```
```bash
SM_CHANNEL_TRAINING='/opt/ml/input/data/training'
SM_CHANNEL_TESTING='/opt/ml/input/data/testing'
```
Contains the directory where the channel named ```channel_name``` is located in the container. Usage examples:

```python
import json

parser.add_argument('--train', type=str, default=os.environ['SM_CHANNEL_TRAINING'])
parser.add_argument('--test', type=str, default=os.environ['SM_CHANNEL_TESTING'])

    
args = parser.parse_args()

train_file = np.load(os.path.join(args.train, 'train.npz'))
test_file = np.load(os.path.join(args.test, 'test.npz'))
```

### SM_HPS
```bash
SM_HPS='{"batch-size": "256", "learning-rate": "0.0001","communicator": "pure_nccl"}'
```
Contains a JSON encoded dictionary with the user provided hyperparameters. Example usage:

```python
import json

hyperparameters = json.loads(os.environ['SM_HPS']))
# {"batch-size": 256, "learning-rate": 0.0001, "communicator": "pure_nccl"}
```
### SM_HP_```{hyperparameter_name}```
```bash
SM_HP_LEARNING-RATE=0.0001
SM_HP_BATCH-SIZE=10000
SM_HP_COMMUNICATOR=pure_nccl
```
Contains value of the hyperparameter named ```hyperparameter_name```. Usage examples:

```python
learning_rate = float(os.environ['SM_HP_LEARNING-RATE'])
batch_size = int(os.environ['SM_HP_BATCH-SIZE'])
comminicator = os.environ['SM_HP_COMMUNICATOR']
```
#### SM_CURRENT_HOST
```json
SM_CURRENT_HOST=algo-1
```
The name of the current container on the container network. Usage example:

```python
# using it in argparse
parser.add_argument('current_host', type=str, default=os.environ['SM_CURRENT_HOST'])

# using it as variable
current_host = os.environ['SM_CURRENT_HOST']
```

#### SM_HOSTS
```json
SM_HOSTS='["algo-1","algo-2"]'
```
JSON encoded list containing all the hosts . Usage example:

```python
import json

# using it in argparse
parser.add_argument('hosts', type=nargs, default=json.loads(os.environ['SM_HOSTS']))

# using it as variable
hosts = json.loads(os.environ['SM_HOSTS'])
```

#### SM_NUM_GPUS
```json
SM_NUM_GPUS=1
```
The number of gpus available in the current container. Usage example:

```python
# using it in argparse
parser.add_argument('num_gpus', type=int, default=os.environ['SM_NUM_GPUS'])

# using it as variable
num_gpus = int(os.environ['SM_NUM_GPUS'])
```
## Environment Variables full specification:
#### SM_NUM_CPUS
```json
SM_NUM_CPUS=32
```
The number of cpus available in the current container. Usage example:
```python
# using it in argparse
parser.add_argument('num_cpus', type=int, default=os.environ['SM_NUM_CPUS'])

# using it as variable
num_cpus = int(os.environ['SM_NUM_CPUS'])
```


#### SM_LOG_LEVEL
```json
SM_LOG_LEVEL=20
```

The current log level in the container. Usage example:
```python
import logging

logger = logging.getLogger(__name__)

logger.setLevel(int(os.environ.get('SM_LOG_LEVEL', logging.INFO)))
```

### SM_NETWORK_INTERFACE_NAME
```json
SM_NETWORK_INTERFACE_NAME=ethwe
```
Name of the network interface, useful for distributed training. Usage example:
```python
# using it in argparse
parser.add_argument('network_interface', type=str, default=os.environ['SM_NETWORK_INTERFACE_NAME'])

# using it as variable
network_interface = os.environ['SM_NETWORK_INTERFACE_NAME']
```

### SM_USER_ARGS
```json
SM_USER_ARGS='["--batch-size","256","--learning_rate","0.0001","--communicator","pure_nccl"]'
```

JSON encoded list with the script arguments provided for training.

### SM_INPUT_DIR
```json
SM_INPUT_DIR=/opt/ml/input/
```
The path of the input directory, e.g. ```/opt/ml/input/```
The input_dir, e.g. ```/opt/ml/input/```, is the directory where SageMaker saves input data and configuration files before and during training.

### SM_INPUT_CONFIG_DIR
```json
SM_INPUT_DIR=/opt/ml/input/config
```
The path of the input directory, e.g. ```/opt/ml/input/config/```. The directory where standard SageMaker configuration files are located, e.g. ```/opt/ml/input/config/```.

SageMaker training creates the following files in this folder when training starts:
- `hyperparameters.json`: Amazon SageMaker makes the hyperparameters in a CreateTrainingJob request available in this file.
- `inputdataconfig.json`: You specify data channel information in the InputDataConfig parameter in a CreateTrainingJob request. Amazon SageMaker makes this information available in this file.
- `resourceconfig.json`: name of the current host and all host containers in the training.

More information about this files can be find here:
    https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html

### SM_OUTPUT_DATA_DIR
```json
SM_OUTPUT_DATA_DIR=/opt/ml/output/data/algo-1
```
The dir to write non-model training artifacts (e.g. evaluation results) which will be retained by SageMaker, e.g. ```/opt/ml/output/data```. 

As your algorithm runs in a container, it generates output including the status of the training job and model and output artifacts. Your algorithm should write this information to the this directory.

### SM_RESOURCE_CONFIG
```json
SM_RESOURCE_CONFIG='{"current_host":"algo-1","hosts":["algo-1","algo-2"]}'
```
The contents from ```/opt/ml/input/config/resourceconfig.json```. It has the following keys:
- current_host: The name of the current container on the container network.
    For example, ```'algo-1'```.
-  hosts: The list of names of all containers on the container network,
    sorted lexicographically. For example, `['algo-1', 'algo-2', 'algo-3']`
    for a three-node cluster.

For more information about resourceconfig.json:
https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html#your-algorithms-training-algo-running-container-dist-training

### SM_INPUT_DATA_CONFIG
```json
SM_INPUT_DATA_CONFIG='{
    "testing": {
        "RecordWrapperType": "None",
        "S3DistributionType": "FullyReplicated",
        "TrainingInputMode": "File"
    },
    "training": {
        "RecordWrapperType": "None",
        "S3DistributionType": "FullyReplicated",
        "TrainingInputMode": "File"
    }
}'
```
Input data configuration from ```/opt/ml/input/config/inputdataconfig.json```.

For more information about inpudataconfig.json:
  https://docs.aws.amazon.com/sagemaker/latest/dg/your-algorithms-training-algo.html#your-algorithms-training-algo-running-container-dist-training

### SM_TRAINING_ENV
```python
SM_TRAINING_ENV='
{
    "channel_input_dirs": {
        "test": "/opt/ml/input/data/testing",
        "train": "/opt/ml/input/data/training"
    },
    "current_host": "algo-1",
    "framework_module": "sagemaker_chainer_container.training:main",
    "hosts": [
        "algo-1",
        "algo-2"
    ],
    "hyperparameters": {
        "batch-size": 10000,
        "epochs": 1
    },
    "input_config_dir": "/opt/ml/input/config",
    "input_data_config": {
        "test": {
            "RecordWrapperType": "None",
            "S3DistributionType": "FullyReplicated",
            "TrainingInputMode": "File"
        },
        "train": {
            "RecordWrapperType": "None",
            "S3DistributionType": "FullyReplicated",
            "TrainingInputMode": "File"
        }
    },
    "input_dir": "/opt/ml/input",
    "job_name": "preprod-chainer-2018-05-31-06-27-15-511",
    "log_level": 20,
    "model_dir": "/opt/ml/model",
    "module_dir": "s3://sagemaker-{aws-region}-{aws-id}/{training-job-name}/source/sourcedir.tar.gz",
    "module_name": "user_script",
    "network_interface_name": "ethwe",
    "num_cpus": 4,
    "num_gpus": 1,
    "output_data_dir": "/opt/ml/output/data/algo-1",
    "output_dir": "/opt/ml/output",
    "resource_config": {
        "current_host": "algo-1",
        "hosts": [
            "algo-1",
            "algo-2"
        ]
    }
}'
```
Provides the entire training information as a JSON encoded dictionary.
## License

This library is licensed under the Apache 2.0 License. 


CHANGELOG.md
# Changelog

## v2.8.2 (2020-03-20)

### Bug Fixes and Other Changes

 * extract module to correct location in download_and_install
 * vulnerable dependency version

### Testing and Release Infrastructure

 * create __init__.py file for Python 2 import of protobuf during tests

## v2.8.1 (2020-02-20)

### Bug Fixes and Other Changes

 * update copyright year

## v2.8.0 (2020-02-19)

### Features

 * pin inotify-simple version.

## v2.7.0 (2020-02-17)

### Features

 * support specifying S3 endpoint URL

### Bug Fixes and Other Changes

 * Fix memory leak in gethostname and adapt len semantics to Posix

## v2.6.4 (2020-02-07)

### Bug Fixes and Other Changes

 * modify download_and_install to work with local tarball
 * pin scipy version >= 1.2.2

## v2.6.3 (2020-02-06)

### Bug Fixes and Other Changes

 * patch module.download_and_install to resolve network isolation issue

## v2.6.2 (2019-12-18)

### Bug fixes and other changes

 * Add `scipy` to requried packages

## v2.6.1 (2019-11-30)

### Bug fixes and other changes

 * bug-fix: array_to_recordio_protobuf should return byte buffer instead of Stream
 * bug-fix: Typo in the execution-parameters routing rule

## v2.6.0 (2019-11-25)

### Features

 * adding support for execution_parameters endpoint for serving

## v2.5.12 (2019-11-15)

### Bug fixes and other changes

 * Adding support for encoding to recordio

## v2.5.11 (2019-10-29)

### Bug fixes and other changes

 * stream stderr even when capture_error is True

## v2.5.10 (2019-10-24)

### Bug fixes and other changes

 * use built-in csv library in csv encoding/decoding for correct quoted string handling.

## v2.5.9 (2019-09-25)

### Bug fixes and other changes

 * Patch os.path.exists for sshd

## v2.5.8 (2019-09-24)

### Bug fixes and other changes

 * Mark gethostname tests as xfail if run locally

## v2.5.7 (2019-09-23)

### Bug fixes and other changes

 * Add Pylint to development process

## v2.5.6 (2019-09-19)

### Bug fixes and other changes

 * Use copy when installing user module from local path
 * Integrate black into development process

## v2.5.5 (2019-07-31)

### Bug fixes and other changes

 * Update setup.py

## v2.5.4 (2019-07-30)

### Bug fixes and other changes

 * install user module before GUnicorn starts
 * include /opt/ml/code to GUnicorn PYTHONPATH

## v2.5.3 (2019-07-22)

### Bug fixes and other changes

 * ensure exit code is an int

## v2.5.2 (2019-07-18)

### Bug fixes and other changes

 * pin flake and werkzeug versions
 * add GPU default for MPI processes per host

### Documentation changes

 * fix env var in readme

## v2.5.1 (2019-06-27)

### Bug fixes and other changes

 * Added execution-parameters to nginx.conf.template

## v2.5.0 (2019-06-24)

### Features

 * entrypoint run waits for hostname resolution

## v2.4.10.post0 (2019-05-29)

### Documentation changes

 * fix path for training script location

## v2.4.10 (2019-05-20)

### Bug fixes and other changes

 * Detailed documentation for SageMaker Containers - training
 * download_and_extract local tar file

## v2.4.9 (2019-05-08)

### Bug fixes and other changes

 * add test for network isolation mode training
 * remove unnecessary name argument from download and extract function

## v2.4.8 (2019-05-02)

### Bug fixes and other changes

 * use mpi4py in MPI command for Python executables

## v2.4.7 (2019-04-30)

### Bug fixes and other changes

 * allow MPI options to be passed through entry_point.run

## v2.4.6.post0 (2019-04-24)

### Documentation changes

 * add commit message format to CONTRIBUTING.md and PR template

## v2.4.6 (2019-04-23)

### Bug fixes and other changes

 * update for automated releases

## v2.4.5

* bug-fix: use specified args, entry point, and env vars when creating a runner

## v2.4.4.post2

* doc-fix: Convert README to RST
* doc-fix: Update README with newer frameworks using SageMaker Containers

## v2.4.4.post1

* Specify ``long_description_content_type`` in setup

## v2.4.4

* bug-fix: correctly set NGINX_PROXY_READ_TIMEOUT to match model_sever_timeout.
* enhancement: remove numpy version restriction.

## v2.4.3

* bug-fix: Fix recursive directory navigation in intermediate output.

## v2.4.2

* bug-fix: Rename libchangehostname to gethostname to match POSIX function name

## v2.4.1

* feature: C extension reads hostname from resourceconfig instead of env var.

## v2.4.0

* feature: Generic OpenMPI support
* bug-fix: Fix response content_type handling

## v2.3.5

* bug-fix: Accept header ANY ('*/*') fallback to default accept
* feature: Add intermediate output to S3 during training
* bug-fix: reintroduce ``_modules.s3_download`` and ``_modules.download_and_install`` for backward compatibility

## v2.3.4

* feature: add capture_error flag to process.check_error and process.create and to all functions that runs process: modules.run, modules.run_module, and entry_point.run

## v2.3.3

* bug-fix: reintroduce _modules.prepare to import_module

## v2.3.2

* bug-fix: reintroduce _modules.prepare for backwards compatibility

## v2.3.1

* [breaking change] remove ``_modules.prepare`` and ``_modules.download_and_install``
* [breaking change] move ``_modules.s3_download`` to ``_files.s3_download``
* feature: support for Bash commands and Python scripts

## v2.3.0

* feature: Allow for dynamic nginx.conf creation
* feature: Provide support for additional environment variables. (http_port, safe_port_range and accept)

## v2.2.7

* feature: Making pip install less noisy
* bug-fix: Stream stderr instead of capturing it when running user script

## v2.2.6

* feature: Make it optional for run_module method to wait for the subprocess to exit
* feature: Allow additional sagemaker hyperparameters to be stored in TrainingEnv

## v2.2.5

* feature: Transformer: support user-supplied ``transform_fn``

## v2.2.4

* bug-fix: remove request size limit correctly

## v2.2.3

* enhancement: remove request size limit

## v2.2.2

* bug-fix: Fix choosing region for S3 client

## v2.2.1


src/sagemaker_containers/_mpi.py
# Copyright 2018-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the 'License'). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://aws.amazon.com/apache2.0/
#
# or in the 'license' file accompanying this file. This file is
# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.
"""Placeholder docstring"""
import argparse
import inspect
import logging
import os
import subprocess
import time
from typing import Any, List, Tuple  # noqa ignore=F401 imported but unused

import paramiko
import psutil

import gethostname
from sagemaker_containers import _logging, _process, _timeout

logger = _logging.get_logger()
logging.getLogger("paramiko").setLevel(logging.INFO)


class WorkerRunner(_process.ProcessRunner):
    """Runner responsible for preparing MPI distributed training and waiting for MPI
     master execution to finish.
    """

    def __init__(self, user_entry_point, args, env_vars, master_hostname):
        """Placeholder docstring"""
        super(WorkerRunner, self).__init__(user_entry_point, args, env_vars)
        self._master_hostname = str(master_hostname)

    def run(
        self, wait=True, capture_error=False
    ):  # type: (bool, bool) -> None # pylint: disable=unused-argument
        """The WorkerRunner proceeds as following:

        - wait for the MPI Master to create its SSH daemon
        - start its SSH daemon
        - monitor the MPI orted process and wait it to finish the MPI execution
        """
        logger.info("Starting MPI run as worker node.")
        if wait:
            logger.info("Waiting for MPI Master to create SSH daemon.")
            self._wait_master_to_start()
        logger.info("MPI Master online, creating SSH daemon.")

        _start_sshd_daemon()

        if wait:
            logger.info("Waiting for MPI process to finish.")
            _wait_orted_process_to_finish()
            time.sleep(30)
        logger.info("MPI process finished.")

    def _wait_master_to_start(self):  # type: () -> None
        """Placeholder docstring"""
        while not _can_connect(self._master_hostname):
            time.sleep(1)

    def _wait_master_to_finish(self):  # type: () -> None
        """Placeholder docstring"""
        while _can_connect(self._master_hostname):
            time.sleep(30)


def _wait_orted_process_to_finish():  # type: () -> None
    """Placeholder docstring"""
    orted = _orted_process()
    psutil.wait_procs(orted)


def _orted_process():  # pylint: disable=inconsistent-return-statements
    """Waits maximum of 5 minutes for orted process to start"""
    for _ in range(5 * 60):
        procs = [p for p in psutil.process_iter(attrs=["name"]) if p.info["name"] == "orted"]

        if procs:
            return procs
        time.sleep(1)


class MasterRunner(_process.ProcessRunner):
    """Responsible to prepare MPI distributed training and syncronize work with the Workers.
    """

    def __init__(
        self,
        user_entry_point,
        args,
        env_vars,
        master_hostname,
        hosts,
        process_per_host,
        custom_mpi_options,
        network_interface_name,
        interval=1,
        timeout_in_seconds=60 * 60,
        num_processes=None,
    ):
        """Placeholder docstring"""

        super(MasterRunner, self).__init__(user_entry_point, args, env_vars)

        self._master_hostname = master_hostname
        self._hosts = hosts
        self._process_per_host = process_per_host
        self._num_processes = num_processes
        self._custom_mpi_options = custom_mpi_options
        self._network_interface_name = network_interface_name
        self._interval = interval
        self.timeout_in_seconds = timeout_in_seconds

    def _setup(self):  # type: () -> None
        """Placeholder docstring"""
        logger.info("Starting MPI run as worker node.")
        logger.info("Creating SSH daemon.")
        _start_sshd_daemon()

        self._wait_for_workers()

    def _wait_for_workers(self):  # type: () -> None
        """Placeholder docstring"""
        logger.info("Waiting for MPI workers to establish their SSH connections")

        workers = [host for host in self._hosts if host != self._master_hostname]
        with _timeout.timeout(seconds=self.timeout_in_seconds):
            for host in workers:
                while not _can_connect(host):
                    time.sleep(self._interval)
                logger.info("Worker %s available for communication", host)

    def _create_command(self):  # type: () -> List[str, Any]
        """Placeholder docstring"""
        num_hosts = len(self._hosts)
        num_processes = self._num_processes or self._process_per_host * num_hosts

        # By default, use one process per GPU, or one process per node (if training with CPU).
        if self._process_per_host == 1:
            host_list = self._hosts
        else:
            host_list = ["%s:%s" % (host, self._process_per_host) for host in self._hosts]

        msg = "Env Hosts: %s Hosts: %s process_per_hosts: %s num_processes: %s"
        logger.info(msg, self._hosts, host_list, self._process_per_host, num_processes)

        overridden_known_options, additional_options = _parse_custom_mpi_options(
            self._custom_mpi_options
        )

        logger.info("Network interface name: %s" % self._network_interface_name)

        command = [
            "mpirun",
            "--host",
            ",".join(host_list),
            "-np",
            str(num_processes),
            "--allow-run-as-root",
            "--display-map",
            "--tag-output",
            "-mca",
            "btl_tcp_if_include",
            self._network_interface_name,
            "-mca",
            "oob_tcp_if_include",
            self._network_interface_name,
            "-mca",
            "plm_rsh_no_tree_spawn",
            "1",
            "-bind-to",
            "socket",
            "-map-by",
            "slot",
            "-mca",
            "pml",
            "ob1",
            "-mca",
            "btl",
            "^openib",
            "-mca",
            "orte_abort_on_non_zero_status",
            "1",
            "-x",
            "NCCL_MIN_NRINGS=4",
            "-x",
            "NCCL_SOCKET_IFNAME=%s" % self._network_interface_name,
            "-x",
            "NCCL_DEBUG=%s" % overridden_known_options.NCCL_DEBUG,
            "-x",
            "LD_LIBRARY_PATH",
            "-x",
            "PATH",
            "-x",
            "LD_PRELOAD=%s" % inspect.getfile(gethostname),
        ]

        command.extend(additional_options)

        for credential in ["AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_SESSION_TOKEN"]:
            if credential in os.environ:
                command.extend(["-x", credential])

        for name in self._env_vars:
            command.extend(["-x", name])

        command.extend(super(MasterRunner, self)._create_command())

        return command

    def _python_command(self):
        """Use mpi4py to force processes to abort if an uncaught exception occurs.
        https://docs.chainer.org/en/stable/chainermn/tutorial/tips_faqs.html#mpi-process-hangs-after-an-unhandled-python-exception
        """
        return super(MasterRunner, self)._python_command() + ["-m", "mpi4py"]


_SSH_DAEMON_NOT_FOUND_ERROR_MESSAGE = """
SSH daemon not found, please install SSH to allow MPI to communicate different nodes in cluster.

You can install ssh by running following commands:
-------------------------------------------------

1. Install SSH via apt-get:

apt-get update && apt-get install -y --no-install-recommends openssh-server && mkdir -p /var/run/sshd

2. SSH login fix. Otherwise user is kicked off after login:
sed 's@session\\s*required\\s*pam_loginuid.so@session optional pam_loginuid.so@g' -i /etc/pam.d/sshd

3. Create SSH key to allow password less ssh between different docker instances:
mkdir -p /root/.ssh/ && ssh-keygen -q -t rsa -N '' -f /root/.ssh/id_rsa && \
  cp /root/.ssh/id_rsa.pub /root/.ssh/authorized_keys && \
  printf "Host *\n  StrictHostKeyChecking no\n" >> /root/.ssh/config
"""


def _start_sshd_daemon():  # type: () -> None
    """Placeholder docstring"""
    sshd_executable = "/usr/sbin/sshd"

    if not os.path.exists(sshd_executable):
        raise RuntimeError(_SSH_DAEMON_NOT_FOUND_ERROR_MESSAGE)

    subprocess.Popen([sshd_executable, "-D"])


def _can_connect(host, port=22):  # type: (str, int) -> bool
    """Checks if the connection to provided ``host`` and ``port`` is possible or not.
       Args:
           host (str): Hostname for the host to check connection.
           port (int): Port name of the host to check connection on.
    """
    try:
        logger.debug("Testing connection to host %s", host)
        client = paramiko.SSHClient()
        client.load_system_host_keys()
        client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
        client.connect(host, port=port)
        client.close()
        logger.info("Can connect to host %s", host)
        return True
    except Exception as e:  # pylint: disable=broad-except
        logger.info("Cannot connect to host %s", host)

        logger.info("Connection failed with exception: \n %s", str(e))
        return False


def _parse_custom_mpi_options(custom_mpi_options):
    # type: (str) -> Tuple[argparse.Namespace, List[str]]
    """Parse custom MPI options provided by user. Known options default value will be overridden
    and unknown options would be identified separately."""

    parser = argparse.ArgumentParser()
    parser.add_argument("--NCCL_DEBUG", default="INFO", type=str)

    return parser.parse_known_args(custom_mpi_options.split())
